ipset.go

```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
 	}
 	if len(in) > 0 {
 		// Ran out of removals before the end of in.
-		ret = append(ret, in...)
+		min = append(min, in...)
 		if debug {
-			debugf("ret=%v", ret)
+			debugf("min=%v", min)
 		}
 	}
 
-	// TODO: possibly update s.in and s.out, if #110 supports that.
+	s.in = min
+	s.out = nil
+}
+
+// Add adds ip to the set s.
+func (s *IPSetBuilder) Add(ip IP) { s.AddRange(IPRange{ip, ip}) }
+
+// AddPrefix adds p's range to s.
+func (s *IPSetBuilder) AddPrefix(p IPPrefix) { s.AddRange(p.Range()) }
 
-	return ret
+// AddRange adds r to s.
+func (s *IPSetBuilder) AddRange(r IPRange) {
+	if !r.Valid() {
+		return
+	}
+	// If there are any removals (s.out), then we need to compact the set
+	// first to get the order right.
+	if len(s.out) > 0 {
+		s.normalize()
+	}
+	s.in = append(s.in, r)
+}
+
+// AddSet adds all ranges in b to s.
+func (s *IPSetBuilder) AddSet(b *IPSet) {
+	for _, r := range b.rr {
+		s.AddRange(r)
+	}
+}
+
+// Remove removes ip from the set s.
+func (s *IPSetBuilder) Remove(ip IP) { s.RemoveRange(IPRange{ip, ip}) }
+
+// RemoveFreePrefix removes and returns a Prefix of length bits from the IPSetBuilder.
+func (s *IPSetBuilder) RemoveFreePrefix(bitLen uint8) (p IPPrefix, ok bool) {
+	s.normalize()
+
+	var bestFit IPPrefix
+	for _, r := range s.in {
+		for _, prefix := range r.Prefixes() {
+			if prefix.Bits > bitLen {
+				continue
+			}
+			if bestFit.IP.IsZero() || prefix.Bits > bestFit.Bits {
+				bestFit = prefix
+				if bestFit.Bits == bitLen {
+					// exact match, done.
+					break
+				}
+			}
+		}
+	}
+
+	if bestFit.IP.IsZero() {
+		return IPPrefix{}, false
+	}
+
+	prefix := IPPrefix{IP: bestFit.IP, Bits: bitLen}
+	s.RemovePrefix(prefix)
+	return prefix, true
+}
+
+// RemovePrefix removes p's range from s.
+func (s *IPSetBuilder) RemovePrefix(p IPPrefix) { s.RemoveRange(p.Range()) }
+
+// RemoveRange removes r from s.
+func (s *IPSetBuilder) RemoveRange(r IPRange) {
+	if r.Valid() {
+		s.out = append(s.out, r)
+	}
+}
+
+// RemoveSet removes all ranges in o from s.
```

Comment:
// RemoveSet removes all IPs in o from s.

Also, why `o` here and `b` above? Let's stick with `b`, because it's less visually ambiguous.